### ISSUE TITLE
different plugins for non-master branches?

### ISSUE THREAD

**Issue by shirotech** [semantic-release/semantic-release#1204]
Hi there,

Is it possible to give a different plugins array to non-master branches?

I am using the beta version of semantic-release to support prereleases and it does not make sense to have "changelog" for a prerelease so in a branch like `beta` or `next`, I don't want it writing to the changelog. Was wondering if there's a way to do it?

**Comment by dominykas** [comment]
You could try to detect the branch yourself and return a different config based on that from you .releaserc.js

**Comment by shirotech** [comment]
Cool, I think I'll switch to using `release.config.js` but I'm trying to get the current branch name if it's already available. I tried `console.log(process.env);` and `console.log(process.argv);` but nothing really useful. Don't really want to hack it with `$(git rev-parse --abbrev-ref HEAD)` if there's already a way to get the current branch name?

**Comment by gr2m** [comment]
I think I would fork whatever plugin you use to generate the changelog and put the logic in there. It might be worth contributing back to upstream if others consider it useful, too?

**Comment by shirotech** [comment]
@gr2m Would love to, but it seems the project member is not very keen on it https://github.com/semantic-release/changelog/issues/51

**Comment by gr2m** [comment]
I think Pierre just does not yet understand the use case, try convince him. If you need it anyway, I’d start out by forking and implementing the change, then you can point to your own repository to show how it works and explain why it’s useful?

**Comment by shirotech** [comment]
I can give it a try later this week, but for now, I have a working solution in case anyone needs it urgent or something.

```js
const {GIT_BRANCH: branch} = process.env;

module.exports = {
  plugins: branch === 'master' ? [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/changelog',
    '@semantic-release/npm',
    '@semantic-release/git',
  ] : [
    '@semantic-release/commit-analyzer',
    '@semantic-release/npm',
  ],
};
```

And in package.json just need to override the semantic-release bin/script.

```json
{
  "scripts": {
    "semantic-release": "cross-env GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD) semantic-release"
  }
}
```

**Comment by pvdlg** [comment]
It's important to point out that prerelease are releases like any other. The only difference is the way increment the version number and on which distribution channel we make them available.

Therefore they should have a changelog just as any other releases. It also make sense for users of a prerelease to know what's inside.

We won't implement any feature without a good reason. Maybe there is something I'm missing, so I asked you in semantic-release/changelog#51 to explain why you don't want the changelogs in pre-release.

If we don't want to implement a feature we would say so and explain why. If we ask questions, it's because we want to know if there is a reason to implement a feature that we are missing.
Not providing a justification for a feature and just opening the same issue in multiple repo will not achieve anything.

**Comment by shirotech** [comment]
These are two separate issues, changelog was just an example. For this issue It’s more just giving the flexibility for users to customise which plugins to use for different branches or purposes.

An example would be if we write the changelog for prereleases it gets copied or duplicated when actual release occurs. Furthermore the release commits get really messy as well. So what is the usual workflow then, do people need to go through a painful rebase every release?

**Comment by pvdlg** [comment]
> These are two separate issues, changelog was just an example. For this issue It’s more just giving the flexibility for users to customise which plugins to use for different branches or purposes.

Allowing different plugins setup per branch would add a lot of complexity in the code base and in the configuration. It also open the door to a lot of inconsistencies. I'd rather make changes in individual plugins when it makes sense.

> An example would be if we write the changelog for prereleases it gets copied or duplicated when actual release occurs.
Pre-release are actual releases. When you merge a prerelease branch to a regular branch a new release happens. That new release will contains the commits of the previous pre-releaase and yes those commits will appear twice in the changelog. But that's not really a problem. Users are just interested in the content of the release they are using at the moment and if doesn't really matter if some commits were released before in a pre-release.

> Furthermore the release commits get really messy as well.
Why? The release commits done by semantic-release (via `@semantic-release/git`) contains the `package.json` updated version and the changelog. How is that messy?

> So what is the usual workflow then, do people need to go through a painful rebase every release?
Making a release should not involve a rebase. If you rebase you will rewrite the commit history and you will end up with release tags that point to commits that doesn't exist anymore.
However if you maintain multiple branches in parallel you will have to merge them at some point and resolve potential conflict. You shouldn't have any conflict in the changelog because each release is always unique (you can't release 1.0.0 twice, if you release 1.0.0-beta.1 and then 1.0.0, that's 2 different releases). You will have new entries in the changelog on both branches, but they won't conflict and all those entries should be preserved in the merged changelog.

**Comment by rahdjoudj** [comment]
Hi guys,
I was thinking about changing .releaserc based on the branch i'm on too.
My issue are merge conflict when I wan't to promote Dev to Test for example.
Dev Branch: beta
Test Branch: normal branch (could be master)

The process:
- commit(s) in **Dev** branch
- beta release which generate changelog that get committed.
- Merge into **Test** branch
- release which also generate changelog entries on top of the merged ones.
- Commit(s) to **Dev** branch
- beta release with new Changelog committed.
- Merge into Test branch
  - **Merge Conflict as Test release also modified the Changelog file.**

I'm using
```"semantic-release": "^16.0.0-beta.22",
    "@semantic-release/commit-analyzer": "^7.0.0-beta.1",
    "@semantic-release/git": "^7.1.0-beta.3",
    "@semantic-release/changelog": "^3.0.4",
    "@semantic-release/exec": "^3.4.0-beta.2",
    "semantic-release-ado": "^1.0.2",
```

.releaserc
```
"plugins": [
  "@semantic-release/commit-analyzer",
  "@semantic-release/release-notes-generator",
  [
    "@semantic-release/changelog",
    {
      "changelogFile": "CHANGELOG.md"
    }
  ],
  "semantic-release-ado",
  [
    "@semantic-release/git",
    {
      "assets": [
        "CHANGELOG.md"
      ],
      "message": "chore(release): ${nextRelease.version} [ci skip]"
    }
  ]
]
```

My logic might be wrong ... but I started to think about not doing Changelog commit for beta 
Any idea on this @pvdlg ? :)

**Comment by pvdlg** [comment]
As it was [mentioned above](https://github.com/semantic-release/semantic-release/issues/1204#issuecomment-501492112) the proper technical solution would be to implement that at the plugin level.

However, it doesn't seems anyone in this thread provided a compelling reason to support such feature. The only example provided was to not include a CHANGELOG.md in pre-release but no one explained why it would make sense to do so.

**Comment by mikecbrant** [comment]
+1 on need to address this issue (probably before prerelease functionality is out of beta).

Since adopting use of prerelease, we are finding that default behavior  is problematic as you begin to get prerelease notes intermigled with (duplicate) release notes once the changes are released into master release branch.

This causes us to have to manually correct changelog files after the master release :(

I guess alternately, rather than just having an option to potentially turn of this changelog plug-in for prerelease, the release-notes-generator could be made to modify prerelease changelog entries into master release entries rather than just simply appending the duplicate master release notes into the changelog. My guess is that this would be more complex to implement though.

**Comment by travi** [comment]
while i can maybe understand your point from the perspective of _producing_ the release notes, be sure that you are also considering the _consumer_ of the notes. 

semantic-release currently produces full notes for each pre-release, which is very useful to those users that are trying out your pre-release. this informs them of the level of impact and the breakages that they should expect to deal with. then, when the pre-releases are promoted to the default channel, the release pulls up the notes from all of the pre-releases into a combined set of release notes for the full release. while there is duplication, this is much easier for a consumer to understand when they only subscribe to releases on the default channel, saving them the effort of going through the notes of each pre-release one by one.

tl;dr, i would be frustrated as a consumer if you modified the behavior to be less informative, especially if i was trying to give early feedback on a potentially unstable release line

**Comment by mikecbrant** [comment]
@travi 

I agree that the end consumer of the release notes is what's most important here.  To me, the **best** consumer experience would be to have updated release notes in the prerelease packages and non-duplicated release notes in the `@latest` package.  For public npm packages, the typical consumer is likely only using `@latest` dist-tag and likely does not care about all the change history in prerelease branches. This is just extra noise and a potential point of confusion.

Just skipping changelog modification for prereleases is, IMO, a half-fix at best, though one I (selfishly) would be happy to have vs. current behavior, at least for our use case where most of our npm modules are private and there really is no "consumer" outside of our development team.

I also think you need to consider the principle of least surprise to the developers themselves.  Let's say that the current beta prelease functionality goes into `semantic-release@latest` with functionality the way it currently is.  Then let's say developers who are already using `semantic-release` see this new prerelease functionality and start using it.  What do you think the developer would expect to see in `@latest` changelog after their prerelease changes have been merged?  I know I was surprised to see these duplicate notes.

**Comment by travi** [comment]
> For public npm packages, the typical consumer is likely only using @latest dist-tag and likely does not care about all the change history in prerelease branches. This is just extra noise and a potential point of confusion.

i can understand the point of not wanting to see a full change history, but they do want a full list of bug fixes, new features, and breaking changes. ideally the commit messages should already be a listing of just that, not a noisy change history. thats kind of the foundation of the approach of semantic-release in the first place.

what you may be suggesting is some level of summarization for the full release, but im not understanding a proposal to make that work with the data that is available to semantic-release. without manually modifying the release notes, i dont know how semantic-release can accomplish that goal. also, publishing the initial notes with semantic-release does not prevent a maintainer from later manually modifying the notes, so there still is that option.

i still see this as exactly the information that i would want as a consumer. if there are lots of changes, i would expect to see lots of fixes, features, and breaking changes listed. reducing the size of that list is best handled by making smaller releases, imho.

**Comment by mikecbrant** [comment]
> what you may be suggesting is some level of summarization for the full release,

Not really a summarization, but a de-duplication.  Perhaps it is the changelog plugin where this needs to happen since the release-notes-generator already works as expected when generating notes since last master release tag.

Since the master release already is given the prerelease changelog file as the starting point when the prerelease branch is merged, it might be as simple removing all entries in changelog file back to the last master release tag before pre-pending the new master release notes to the top.

**Comment by pvdlg** [comment]
The pre-release branches are meant to work on future releases of the package. As such the intent is to have a long lived branch (several weeks or months) on which you will make a lot of breaking changes, distribute those to a subset of your users and then release all of those at once on the general channel.

You are using the pre-release branches in a very different way. It seems you are releasing the code of each of your PR. That's why you don't want the changelog on the long run.

From your comment:
> This allows us to create prereleases for any feature branch on a repo, making it really simple for us to allow multiple developers to independently and simultaneously work on new package features on their own feature branches.

How does make a release for each feature branch allow allow multiple developers to independently and simultaneously work on new package features? 
What allow that is the fact you have multiple branches, not the fact that you are making a release for each of those.

What are you doing with those pre-release? Why do you need them to be published?
A pre-release is meant for being tested by a subset of users. It seems those pre-release are so short lived that it's not the case.

I would recommend to use a traditional [GitHub Workflow](https://guides.github.com/introduction/flow/):
- Release from `master` with semantic-release
- Continue to create feature branches so your developer can work in parallel
-  On each feature branch, develop one feature (that might contain multiple commits, and can be rebased or squashed, etc...)
- Open a PR and have your CI running test
- Merge the PR when the feature is ready, which will trigger semantic-release to release from `master`

**Comment by mikecbrant** [comment]
> You are using the pre-release branches in a very different way. It seems you are releasing the code of each of your PR. That's why you don't want the changelog on the long run.

This is correct.  This is a trunk-based development model as opposed to gitflow or similar where there are long-lived branches.

> How does make a release for each feature branch allow allow multiple developers to independently and simultaneously work on new package features?

I'll start with the standard proposition that an npm package is typically intended to be used by some consuming application(s).  During the course of making changes in such applications, you will often find the need to make a change in dependent package in order to enable the desired functionality.  By having the ability to prerelease on any branch, a developer can simply go to the dependency repo,   create a branch for the change and be able to publish a prerelease package for the specific feature without having to wait on the pull request / review / master release cycle for that dependent package before being able to consume it in the application, and without potentially introducing breakage to other users as one would on a long-running `next`-type branch. 

Additionally, since that prerelease package is available via npm, we have access to it in all of our potential CI environments - local, Bitbucket, AWS Codebuild, etc. such that we do not need to rely on inadequate or problematic solutions like `npm link` or manual `npm pack` in order to work with these new changes.

The workflow looks like:
 - make changes in dependency repo feature branch
 - when passing all test gates, the prerelease package is published
 - open PR for change in the dependency
 - in consuming application, change `package.json` to refer to prerelease package version
 - work on changes in application as you would normally
 - after PR / merge of  dependent package changes, update `package.json` to reflect new `@latest` release
 - perform PR / merge on consuming application

Because prereleases are made on a per-branch basis, a developer does not need to worry about stepping on the toes of other developers who may be working on other changes in the same dependency packages or coordinating their changes with other developers changes on a `next`-style branch.  The developer has the ability to isolate changes in the full dependency tree as relates to their specific change in the consuming application without disrupting any other developer's work.

If you want to have multiple developers working on the same branch (for example. a branch tracking an upcoming major release) you always have the option to do so if you want.  You have the flexibility to use the branches as appropriate to the work or application at hand as opposed to having to subscribe to a one-branching-strategy-fits-all approach.

If you took your suggested approach on only releasing `master` (the way we used to do it until we discovered this great prerelease functionality), you end up having developers blocked waiting for PR / merge / release cycle and you can't use modified packages in remote CI environments without provisioning some means for getting the candidate package tarball into the CI environment.

By publishing these prerelease versions, everything "just works" in all CI environments with only changes to `package.json` needed in the consuming app to utilize the prerelease packages.

**Comment by travi** [comment]
> This is a trunk-based development model as opposed to gitflow or similar where there are long-lived branches.

ah, ok. i missed this detail. i agree that this isn't really a pre-release model, but i can see how you might want to make it fit.

> PR / merge / release cycle

you mention trunk-based development, but then you also mention changes being blocked by the PR gate. while the PR gate is common with gitflow/github flow, etc., it is usually discouraged with [trunk-based development](https://trunkbaseddevelopment.com). since semantic-release very directly encourages continuous-deployment, trunk-based development since it encourages continuously integrating changes into `master` for deployment/publishing. 

my team is very much on the trunk-based development approach, but the key difference that we do from what it sounds like is your process is that we push directly to `master` rather than submitting PRs that are blocked by review. we still do reviews, but it is after the code is already in master since we are confident that our tests prove that the behavior at least _works correctly_ and the review is focused on making the code better in areas that aren't specifically about the behavior. 

is removing the PR gate from your process an option at all? if so, would that remove your need to do isolated publishes in this pre-release-like way? it sounds to me like it could greatly simplify your process.

**Comment by travi** [comment]
if it helps paint a more full picture, my team does use pre-releases a fair amount but only for cases where we are introducing known breaking changes that we need to be able to update in consumers before publishing as a full release. that way the consumers actually use the pre-release temporarily in production. once all consumers are updated to be compatible with the pre-release, we promote to `latest`.

**Comment by mikecbrant** [comment]
> is removing the PR gate from your process an option at all? 

I actually don't want this to be an option, as I desire all change to be review before going to master. As PR can pick up subtle design issues, etc, that automated tests may not reveal  :)  

Perhaps the difference in our setups is that we run full CI testing on all of our branches.  In our typical development workflow, we run basic build tests as a prerequisite for even being able to push feature branches.  Our remote  CI environments then perform the same build tests plus any other automated testing appropriate to the package before triggering semantic-release to get prerelease. Similarly, we perform these same tests locally on merged master before pushing the merge to the remote repo.

**Comment by mikecbrant** [comment]
> if it helps paint a more full picture, my team does use pre-releases a fair amount but only for cases where we are introducing known breaking changes that we need to be able to update in consumers before publishing as a full release. that way the consumers actually use the pre-release temporarily in production. once all consumers are updated to be compatible with the pre-release, we promote to latest.

I honestly don't think our usage differs that much in this regards. We do still use prerelease for the "traditional" prerelease case you mention here.  We just don't limit our usage to this specific use case or to special predefined branch names.  We don't care if the branchname is `next`, `alpha`, `beta`, `JIRA-123`, `somefeature`, etc.  Anything that is not `master` can get a prerelease, allowing us to use this functionality to address broader use cases.  For example, on public npm packages we will often use `next` branch to track upcoming breaking change, but if I want to work on some change (and consume it in an application on) some that is intended to land in `next`, but is not ready to be published to `next`, I can branch from `next` to get a different prerelease package and then merge that back into `next` when feature is done.

**Comment by travi** [comment]
> I actually don't want this to be an option, as I desire all change to be review before going to master. As PR can pick up subtle design issues, etc, that automated tests may not reveal :)

if its not an option, its not an option, but at that point you arent actually doing a trunk-based development flow either. if it truly is just things like "subtle design issues" i think its worth considering how risky those really are and if it fully justifies the extra complexity to the process. i'm an advocate of the "make it work, make it right, make it fast" mantra. if it works and that is proven by tests, make it right can happen after merging, either because of a review or just further iteration. your interpretation of "subtle design issues" might be different than mine, but i tend to think they don't impact whether it works correctly or not.

> I honestly don't think our usage differs that much in this regards.

i dont know your flow well enough to comment all that intelligently, but the main difference i see is the extra complexity that you've chosen to introduce by not merging to master ends up causing more release notes than you need. i was only trying to suggest removing the need to publish as pre-releases so that normal publishes wouldnt result in duplicate notes. it still comes down to you leveraging a "pre-release" to accomplish more of a "test build", which isnt directly supported as a feature in the beta.

it truly seems to me that what youre really asking for is less about running different plugin configs for non-master branches, like this issue was originally focused. while enabling that could get you closer to your goal, it seems like youre really after more of a test-build feature in addition to pre-releases and maintenance releases

**Comment by mikecbrant** [comment]
> it truly seems to me that what youre really asking for is less about running different plugin configs for non-master branches.

I would say this is correct.  I originally stumbled across this issue in trying to determine if there was some means to configure changelog output behavior.  I don't care so much about branch-specific plugin configuration (though this could certainly be a way to approach this particular concern and might have other applicable uses).

In going through this conversation, I am guessing what I am looking for might best be captured as a feature request on changelog plugin - who knows perhaps one I eventually submit PR for if manual changelog changes on our public packages ends up being to much manual work. :)

>  it still comes down to you leveraging a "pre-release" to accomplish more of a "test build", which isnt directly supported as a feature in the beta.

Correct in that for our typical use case, we just want to be able to publish candidate packages to npm such that we can develop against them more easily (we do still need the "prerelease" usage as well though from time to time).

The only thing to note is that from an npm standpoint there really is no concept of "prerelease" vs. "test" - there are just dist-tags. It is really only convention in how dist-tags are used to be used as "prerelease".  There are certainly other use cases for dist-tags in addition to being used as "prerelease".  For example, a browser-side library might use different dist-tags for different intended target browsers `@ie`, `@chrome`, etc.

The good news is that the semantic-release beta prerelease functionality can handle these broader  use cases if one wants to use it that way.  Now, whether modifying the changelog creation logic to support these use cases is considered an important feature in the overall scheme of desired semantic-release features, is certainly another question. My guess is that there are many more important features that can be implemented from the general user base standpoint than dealing with this one narrow feature request. :)

At any rate, I didn't mean to take this thread off the original topic.  Hopefully the library maintainers can at least get a window into this particular way in which this library is being used.

**Comment by travi** [comment]
> The only thing to note is that from an npm standpoint there really is no concept of "prerelease" vs. "test" - there are just dist-tags.

i think this is probably the key piece to the conversation. while there isn't a difference at the npm level, there is at the semantic-release level. conventional uses of dist-tags enable defining workflows that simplify the use of those conventions. the way i understand the implementation of the pre-release and maintenance release logic in the beta is along these lines (@pvdlg, please clarify if i'm describing incorrectly). while it may be possible to bend these intended conventions to get closer to support other conventions, i dont think thats directly accounted for in semantic-release. 

> There are certainly other use cases for dist-tags in addition to being used as "prerelease".

agreed, but my understanding is that pre-release and maintenance release are the currently supported cases in this beta. i think this discussion highlighted this particular case well enough to be understood. it probably comes down to whether it is desirable for the conventions to be well defined for common flows and potentially define more flows or be flexible to support "variations" beyond the well defined flows.

**Comment by pvdlg** [comment]
The workflow you describe is not really something we had in mind when implementing the pre-release branches, but I see how it can be useful.

It seems your main reason for publishing each feature branch is dependency management, such that you can can use the code from the feature branch in a package that depends on it.

We actually have a similar situation in the semantic-release repo, where some features in the core require a feature in the plugins. To solve that we define dependencies [with Git branch URLs](https://medium.com/@jonchurch/use-github-branch-as-dependency-in-package-json-5eb609c81f1a). Is that something you considered?

**Comment by mikecbrant** [comment]
> We actually have a similar situation in the semantic-release repo, where some features in the core require a feature in the plugins. To solve that we define dependencies with Git branch URLs. Is that something you considered?

Didn't really consider this as that would seem to break any npm update functionality, and potentially leave composing apps in broken state if the branch is merged/deleted (whereas the npm package would remain).

**Comment by pvdlg** [comment]
When you publish a pre-release and you reference it when working on a feature on a package that depends on it, how do you ensure this doesn't get merged to master?

For example with a package A that depends on package B. On package B you create a feature branch and make a release `1.0.0-my-feature.1`. Then on package A you create a new feature and modify your `package.json` to use the `1.0.0-my-feature.1` version of B.
When you end up merging the feature branch on B to `master` the `package.json` will still reference `1.0.0-my-feature.1` rather than a non-prerelease version.
How do you deal with that? Do you keep some non-prerelease version depending on pre-release one? Or do you fix it somehow in your workflow?

**Comment by travi** [comment]
at least in my case, greenkeeper covers this situation. greenkeeper only sends PRs for out-of-range releases on the `latest` channel, so i have to manually update to depend on the pre-releases, which is desired. when the pre-release is merged to `master`, the new full version is published to `latest` and is out-of-range, so the greenkeeper PR gets the consumer onto the new full version.

not sure the differences with the candidate-build flow described here, but maybe serves as an example to compare to.

**Comment by mikecbrant** [comment]
> It seems your main reason for publishing each feature branch is dependency management, such that you can can use the code from the feature branch in a package that depends on it.

This is true with regards to npm-published packages.  Interestingly, we are also likely to use prerelease pretty soon in some additional (probably non-standard) ways.  For example, we are already using semantic-release master branch tagging as part of our workflow for deploying to AWS - ECS, Lambda, etc. where we give our developer per-branch sandbox environments.

We primarily use it to tag master releases, but I would like to extend that to tag pre-prod releases into sandbox environments as well.  This is not a dependency management use case, but rather a case of using semantic-release to get authoritative release tags, that are more human readable/understandable than just some build id generated by the build environment.  This allows us to associate bugs to a specific versioned release tags (even in pre-prod releases) so as to get better understanding what release(s) a bug may live in or be introduced in.  Ultimately we may use it for API versioning (at least for breaking changes).

> When you publish a pre-release and you reference it when working on a feature on a package that depends on it, how do you ensure this doesn't get merged to master?

So one of the interesting behaviors of npm is that that npm update can take care of the conversion for you.  Say for example the composing app has `^1.1.0-somefeature.1` for its version specifier and the feature has since been merged and `1.1.0@latest` becomes available.  Running `npm update` for this application will update the specifier to `^1.1.0` automatically.

One of the key tenant of our devops process is a strong bias toward regular dependency updates.  We execute `npm update` as part of our automation prior to remote git pushes/merges. So generally speaking, the issue takes care of itself.    Of course a developer could attempt to merge the application change to master before the dependency change has been published to `@latest`.  In this case at least the application won't break (in that the prerelease dependency is still available), but tis would be considered outside of our standard practice.   We also have our pre-master merge code review, so typically the developer would be expected to have updated the `package.json` entry to use the updated dependency prior to review approval / merge. But even if it slips through code review, the devops tooling should correct the dependency prior to merge as I noted.

**Comment by pvdlg** [comment]
Thanks for all the details answer.
The workflow you describe is not really a use case we though of, but it quite interesting to see people using semantic-release in clever ways.

Because it's not a use case we though about from the get go I'm wondering what kind of issues we might get in the future and would like to make sure that the change we would have to make won't creep up in the future.

It seems currently your only issue is to skip the `CHANGELOG.md` for certain branches. Is this correct? Is there any other things that would need to branch specific?

**Comment by mikecbrant** [comment]
> It seems currently your only issue is to skip the CHANGELOG.md for certain branches. Is this correct? Is there any other things that would need to branch specific?

Skipping changelog modification in prerelease branches is certainly one approach to take, though I agree with your thoughts from earlier in the thread that prerelease package should get updated release notes as well.

I tried to capture thoughts on what changes may be needed in changelog package in https://github.com/semantic-release/changelog/issues/62. That might be best place to capture any additional conversation on requirements unless you think that changes at the semantic-release package level might be required (like if the approach of skipping changelog generation based on config might be a better approach in your opinion).   After looking through the changelog plugin codebase, it doesn't seem like it would be that complex to modify `/lib/prepare.js` to (when configured) parse through the release note content and remove what would be duplicate entries caused by merges.

If that library is open to contributors, I might try to squeeze in some time to make the change in the coming weeks if the approach makes sense.

**Comment by pvdlg** [comment]
See https://github.com/semantic-release/changelog/issues/62#issuecomment-535624068
Editing changelogs after the fact is neither a good nor feasible in my opinion.

the only viable option is to add a `filterBranches` in `@semantic-release/changelog` to skip the update of `CHANGELOG.md` for certains branches. Or maybe more simply `updateOnPrerelease` that would be `true` by default.

I just don't know if that would be sufficient.

**Comment by shirotech** [comment]
Probably can close this if the feature request is not aligned with plugin author's vision. I've migrated to https://github.com/conventional-changelog/standard-version which seems to support our needs and it's working great.
